{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "799954\n",
      "984143\n",
      "127974\n",
      "19\n",
      "90\n",
      "79\n"
     ]
    }
   ],
   "source": [
    "# === Algorithms for DNA Sequencing ====\n",
    "# On Coursera, provided by John Hopkins School of Medicine\n",
    "# Programming Homework 2\n",
    "\n",
    "# Provides by staff:\n",
    "import bm_preproc\n",
    "import kmerindex\n",
    "import subseq\n",
    "# Binary Search module\n",
    "import bisect\n",
    "\n",
    "\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "\n",
    "# provided in the problem statment:\n",
    "# Do Boyer-Moore matching. p=pattern, t=text, p_bm=BoyerMoore object for p\n",
    "\n",
    "def boyer_moore(p, p_bm, t):\n",
    "    i = 0\n",
    "    occurrences = []\n",
    "    comparisons = 0\n",
    "    alignments = 0\n",
    "    while i < len(t) - len(p) + 1:\n",
    "        alignments += 1\n",
    "        shift = 1\n",
    "        mismatched = False\n",
    "        for j in range(len(p)-1, -1, -1):\n",
    "            comparisons += 1\n",
    "            if p[j] != t[i+j]:\n",
    "                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n",
    "                skip_gs = p_bm.good_suffix_rule(j)\n",
    "                shift = max(shift, skip_bc, skip_gs)\n",
    "                mismatched = True\n",
    "                break\n",
    "        if not mismatched:\n",
    "            occurrences.append(i)\n",
    "            skip_gs = p_bm.match_skip()\n",
    "            shift = max(shift, skip_gs)\n",
    "        i += shift\n",
    "    return occurrences, comparisons, alignments\n",
    "    \n",
    "    \n",
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    comparisons = 0\n",
    "    alignments = 0\n",
    "    for i in range(len(t) - len(p) + 1):  \n",
    "        alignments += 1\n",
    "        match = True\n",
    "        for j in range(len(p)):  \n",
    "            comparisons += 1\n",
    "            if t[i+j] != p[j]: \n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)\n",
    "    return occurrences, comparisons, alignments\n",
    "\n",
    "def approximate_match(p, t, n):\n",
    "    segment_length = int(round(len(p) / (n+1)))\n",
    "    all_matches = set()\n",
    "    p_idx = kmerindex.Index(t, segment_length)\n",
    "    idx_hits = 0\n",
    "    for i in range(n+1):\n",
    "        start = i*segment_length\n",
    "        end = min((i+1)*segment_length, len(p))\n",
    "        matches = p_idx.query(p[start:end])\n",
    "        \n",
    "        for m in matches:\n",
    "            idx_hits += 1\n",
    "            if m < start or m-start+len(p) > len(t):\n",
    "                continue\n",
    "            \n",
    "            mismatches = 0\n",
    "            \n",
    "            for j in range(0, start):\n",
    "                if not p[j] == t[m-start+j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > n:\n",
    "                        break\n",
    "            for j in range(end, len(p)):\n",
    "                if not p[j] == t[m-start+j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > n:\n",
    "                        break\n",
    "            \n",
    "            if mismatches <= n:\n",
    "                all_matches.add(m - start)\n",
    "    return list(all_matches), idx_hits\n",
    "\n",
    "\n",
    "def approximate_match_subseq(p, t, n, ival):\n",
    "    segment_length = int(round(len(p) / (n+1)))\n",
    "    all_matches = set()\n",
    "    p_idx = subseq.SubseqIndex(t, segment_length, ival)\n",
    "    idx_hits = 0\n",
    "    for i in range(n+1):\n",
    "        start = i\n",
    "        matches = p_idx.query(p[start:])\n",
    "        \n",
    "        for m in matches:\n",
    "            idx_hits += 1\n",
    "            if m < start or m-start+len(p) > len(t):\n",
    "                continue\n",
    "            \n",
    "            mismatches = 0\n",
    "            \n",
    "            for j in range(0, len(p)):\n",
    "                if not p[j] == t[m-start+j]:\n",
    "                    mismatches += 1\n",
    "                    if mismatches > n:\n",
    "                        break\n",
    "            \n",
    "            if mismatches <= n:\n",
    "                all_matches.add(m - start)\n",
    "    return list(all_matches), idx_hits\n",
    "\n",
    "\n",
    "\n",
    "def main():\n",
    "    chr1 = readGenome('chr1.GRCh38.excerpt (1).fasta')\n",
    "    p = 'GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG'\n",
    "    p_bm = bm_preproc.BoyerMoore(p)\n",
    "    # Question 1\n",
    "    print(naive(p, chr1)[2])\n",
    "    # Question 2\n",
    "    print(naive(p, chr1)[1])\n",
    "    # Question 3\n",
    "    print(boyer_moore(p, p_bm, chr1)[2])\n",
    "    p = 'GGCGCGGTGGCTCACGCCTGTAAT'\n",
    "    # Question 4\n",
    "    print(len(approximate_match(p, chr1, 2)[0]))\n",
    "    # Question 5\n",
    "    print(approximate_match(p, chr1, 2)[1])\n",
    "    # Question 6\n",
    "    print(approximate_match_subseq(p, chr1, 2, 3)[1])\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
